File: app/src/main/java/com/marcrobito/amtest/modules/java/MainActivity.java
package com.marcrobito.amtest.modules.java;

import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.ProgressBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.marcrobito.amtest.AppJava;
import com.marcrobito.amtest.R;
import com.marcrobito.amtest.pojos.MovieData;
import com.marcrobito.amtest.utils.MovieJRecyclerViewAdapter;
import com.marcrobito.amtest.utils.OnItemClickListenerJ;

import javax.inject.Inject;
import java.util.List;

public class MainActivity extends AppCompatActivity implements MainJContract.View, OnItemClickListenerJ, BackToMainInterfaceJ {

    @Inject
    public MainJContract.Presenter presenter;

    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;

    @BindView(R.id.fragment_container)
    View container;


    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    List<MovieData> movieData;
    DetailFragment detailFragment;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ((AppJava)getApplication()).getComponent().inject(this);
        setContentView(R.layout.activity_main);

        detailFragment = new DetailFragment();
        getSupportFragmentManager().beginTransaction().add(R.id.fragment_container, detailFragment).commit();
        ButterKnife.bind(this);



    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.setView(this);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount() == 0
        ) {
            onBackPressed();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public void onBackPressed() {
        if(container.getVisibility() == View.VISIBLE){
            container.setVisibility(View.GONE);
        }
    }

    @Override
    public void presentTopMovies(List<MovieData> movieList) {
        this.movieData = movieList;

        MovieJRecyclerViewAdapter adapter = new MovieJRecyclerViewAdapter(movieData, this);
        LinearLayoutManager manager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(manager);
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void onItemClick(Integer position) {
        container.setVisibility(View.VISIBLE);
        detailFragment.setMovie(movieData.get(position));
    }

    @Override
    public void backToMain() {
        onBackPressed();
    }


}

File: app/src/main/java/com/marcrobito/amtest/modules/MainActivity.kt
package com.marcrobito.amtest.modules

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.marcrobito.amtest.R
import com.marcrobito.amtest.pojos.MovieData
import com.marcrobito.amtest.utils.MovieRecyclerViewAdapter
import com.marcrobito.amtest.utils.OnItemClickListener
import kotlinx.android.synthetic.main.activity_main.*
import org.koin.android.ext.android.inject
import android.view.KeyEvent.KEYCODE_BACK
import android.view.KeyEvent
import android.content.Intent




class MainActivity : AppCompatActivity(), MainContract.View, OnItemClickListener, BackToMainInterface {

    private val presenter: MainContract.Presenter by inject()

    private var  movieList:List<MovieData> = listOf()
    private var movieDetailFragment: MovieDetailFragment? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        movieDetailFragment = MovieDetailFragment()

        supportFragmentManager.beginTransaction().add(R.id.fragment_container, movieDetailFragment!!).commit()

    }

    override fun onResume() {
        super.onResume()
        presenter.setView(this)
    }

    override fun onItemClick(position: Int) {
        Log.d("tag",movieList[position].title)
        if(movieDetailFragment != null){
            fragment_container.visibility = View.VISIBLE
            movieDetailFragment!!.setMovie(movieList[position])
        }
    }



    override fun presentTopMovies(movieList: List<MovieData>) {

        this.movieList = movieList

        val adapter = MovieRecyclerViewAdapter(this.movieList, this)
        val manager = LinearLayoutManager(this@MainActivity, LinearLayoutManager.HORIZONTAL, false)

        recyclerView.adapter = adapter
        recyclerView.layoutManager = manager

        progressBar.visibility = View.GONE




    }

    override fun onBackPressed() {
        if(fragment_container.visibility == View.VISIBLE)
            fragment_container.visibility = View.GONE

    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KEYCODE_BACK && event.repeatCount == 0
        ) {
            Log.d("CDA", "onKeyDown Called")
            onBackPressed()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun backToMAin() {
        onBackPressed()
    }
}
